src/libstore/nar-info-disk-cache.cc

```diff
@@ -189,23 +216,42 @@ class NarInfoDiskCacheImpl : public NarInfoDiskCache
             // the cache for this URI in the meantime.
             auto cache(queryCacheRaw(*state, uri));
 
-            if (cache)
+            if (cache && cacheIsUpToDate(*cache))
                 return;
 
-            state->insertCache.use()(uri)(time(0))(storeDir)(wantMassQuery)(priority).exec();
-            assert(sqlite3_changes(state->db) == 1);
-            state->caches[uri] = Cache{(int) sqlite3_last_insert_rowid(state->db), storeDir, wantMassQuery, priority};
+            Cache ret {
+                // id: to be set in the conditional below
+                .id = -1,
+                // lastCacheInfoFetch: always the current time, because this is
+                // only called when a cached Cache wasn't available, so the parameters
+                // are always fresh.
+                .lastCacheInfoFetch = time(0),
+                .storeDir = storeDir,
+                .wantMassQuery = wantMassQuery,
+                .priority = priority,
+            };
+            if (cache) {
+                ret.id = cache->id;
+                state->updateCache.use()(uri)(ret.lastCacheInfoFetch)(storeDir)(wantMassQuery)(priority)(ret.id).exec();
+                assert(sqlite3_changes(state->db) == 1);
+            } else {
+                state->insertCache.use()(uri)(ret.lastCacheInfoFetch)(storeDir)(wantMassQuery)(priority).exec();
+                assert(sqlite3_changes(state->db) == 1);
+                ret.id = (int) sqlite3_last_insert_rowid(state->db);
```

Comment:
```suggestion
                ret.id = (int) state->db.getLastInsertedRowId();
```

(mostly cosmetic, but since we have a more idiomatic way of doing that, we might as well use it)